hub/swap_data_directories.go

```
@@ -0,0 +1,81 @@
+package hub
+
+import (
+	"fmt"
+
+	"github.com/greenplum-db/gp-common-go-libs/gplog"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/utils"
+)
+
+func SwapDataDirectories(hub Hub, agentBroker AgentBroker) error {
+	swapper := finalizer{agentBroker: agentBroker}
+	swapper.archive(hub.masterPair.source)
+	swapper.publish(hub.masterPair.target, hub.masterPair.source)
+	swapper.swapOnAgents(hub.agents)
+	return swapper.Errors()
+}
+
+type finalizer struct {
+	err         *multierror.Error
+	agentBroker AgentBroker
+}
```

I don't understand this. Why can't we just keep it simple with an inline multierror? We should fail fast and bubble up errors like our other idioms, and if not we should have a story that addresses that across our code base in a conscious effort.

---

This struct has been removed.